### PR TITLE
Fixes TLS Mode Defaulting

### PR DESCRIPTION
**Pull request by danehans**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Properly sets the "Terminate" default for `gateway.spec.tls.mode`.

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/service-apis/issues/517

**Does this PR introduce a user-facing change?**:
Yes, sets the default TLS mode to "Terminate".

### CODE REVIEW

**Comment by robscott** [comment]
Thanks! This LGTM other than all the boilerplate year changes. That should be resolved once https://github.com/kubernetes-sigs/service-apis/pull/519 is in.

**Review by hbagdi — APPROVED** [review]
LGTM. This will need a rebase once #519 is merged.

**Comment by hbagdi** [comment]
/hold

Just to make sure this doesn't get merged.

**Comment by hbagdi** [comment]
@danehans Can you rebase to resolve the conflict?
/hold cancel

**Comment by robscott** [comment]
@danehans Thanks for making this PR! After a rebase this LGTM. One small nit - I think this is a user-facing change since we're adding a default value.

**Comment by danehans** [comment]
@robscott I updated the PR description to include user-facing changes.

File: apis/v1alpha1/gateway_types.go
Comment (hbagdi) [inline apis/v1alpha1/gateway_types.go:330]: I was about lgtm this when I recalled why this was *not* the case earlier:
We wanted the user to explicitly set the condition to `Terminate` and make them supply the Certificate in that case.
In this case, if this default kicks in, and a CertificateRef is absent, then how do we handle this case?
We could add validation in the webhook for this case.

Do we still want this to default though?

Comment (robscott) [reply apis/v1alpha1/gateway_types.go:330]: That's a very good point - thanks for catching this! I think that approach makes sense, maybe instead of setting a default here we should clarify that both a CertificateRef and mode must be set as part of TLS config?

Comment (danehans) [reply apis/v1alpha1/gateway_types.go:330]: @hbagdi @robscott if ^ is the case, why not default `Mode` to "Passthrough" so `CertificateRef` can continue to be optional?

Comment (hbagdi) [reply apis/v1alpha1/gateway_types.go:329]: Because `Passthrough` is the rare case. It only makes sense with TCPRoute and TLSRoute (sni sniffing).

Comment (danehans) [reply apis/v1alpha1/gateway_types.go:329]: If `Mode` is optional, then a user can supply a `CertificateRef ` without setting `Mode`, resulting in an unusable TLS config. IMO, if "Terminate" is the common use-case, then it should be the default and the details documented. @hbagdi @robscott @bowei thoughts?

Comment (hbagdi) [reply apis/v1alpha1/gateway_types.go]: The default of `Terminate` makes sense here. What should be the error condition that should be set when CertificateRef is not supplied?

Comment (robscott) [reply apis/v1alpha1/gateway_types.go]: Just to make sure I'm understanding this correctly, this default will not come into play unless someone specifies TLS config. It will essentially allow a shortcut where most users would only need to specify CertificateRef here. Does that sound right? If so, I think this is a very reasonable default.

@hbagdi That's a great question. I think our existing `InvalidCertificateRef` condition would mostly apply here, even if the initial intent was slightly different.

Comment (hbagdi) [reply apis/v1alpha1/gateway_types.go]: > Just to make sure I'm understanding this correctly, this default will not come into play unless someone specifies TLS config. It will essentially allow a shortcut where most users would only need to specify CertificateRef here. Does that sound right?

That's correct. I pulled the patch down and tested it myself.
We actually missed specifying `mode` even in our example: https://github.com/kubernetes-sigs/service-apis/blob/27c4d2dcd3c70a65a1d512fc7a21d2ca4403d218/examples/tls-basic.yaml#L23-L29

> @hbagdi That's a great question. I think our existing InvalidCertificateRef condition would mostly apply here, even if the initial intent was slightly different.

Sounds fair. If the overhead is not too much, I'd advocate for a new AbsentCertificateRef or a generic AbsentRef condition.

**Comment by hbagdi** [comment]
LGTM.
Leaving final approval to Rob.

**Pull request merged**
